### PR TITLE
For #8415 - Speculative RecentlyClosedMiddlewareTest fix

### PR DESCRIPTION
**Pull request by ekager**
---

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by ekager** [comment]
bors try

**Comment by ekager** [comment]
bors try

**Comment by ekager** [comment]
bors try

**Comment by ekager** [comment]
I'm not sure why this would fix it but I'm just trying to get it as close to `ContainerMiddlewareStorageTest` as possible 🤔

**Comment by ekager** [comment]
bors try

File: components/feature/recentlyclosed/src/test/java/mozilla/components/feature/recentlyclosed/RecentlyClosedMiddlewareTest.kt [hunk 1]
```diff
@@ -82,8 +82,6 @@ class RecentlyClosedMiddlewareTest {
             store.dispatch(RecentlyClosedAction.AddClosedTabsAction(listOf(closedTab)))
                 .joinBlocking()
 
-            dispatcher.advanceUntilIdle()
```
Comment (pocmo) [inline components/feature/recentlyclosed/src/test/java/mozilla/components/feature/recentlyclosed/RecentlyClosedMiddlewareTest.kt:85]: I think those calls made sense? In the middleware you are dispatching coroutines on the scope using that dispatcher. So you may want to wait for the to complete before you can verify that something happened after they executed. 🤔 

However some of the coroutines also dispatch something on the store. So you may also want to wait until the store has processed that (which also happens async). 

We added `Store.waitUntilIdle()` for doing that in tests.

**Comment by ekager** [comment]
bors try

File: components/feature/recentlyclosed/src/test/java/mozilla/components/feature/recentlyclosed/RecentlyClosedMiddlewareTest.kt [hunk 2]
```diff
@@ -109,6 +111,7 @@ class RecentlyClosedMiddlewareTest {
 
             store.dispatch(TabListAction.RemoveTabAction("1234")).joinBlocking()
 
+            store.waitUntilIdle()
             dispatcher.advanceUntilIdle()
```
Comment (pocmo) [inline components/feature/recentlyclosed/src/test/java/mozilla/components/feature/recentlyclosed/RecentlyClosedMiddlewareTest.kt:115]: I think we need to flip the order here: Fist we need to wait for the middleware coroutine to complete and dispatch on the store, then, when we know the store will already know about the action, wait for the store.

**Review by pocmo — APPROVED** [review]
👍

**Pull request merged**